frontend/packages/core/src/AppLayout/shortLinker.tsx

```diff
@@ -0,0 +1,193 @@
+
+import { Button, ClipboardButton, IconButton } from "../button";
+import { useAppContext, useShortLinkContext } from "../Contexts";
+import type { HydratedData } from "../Contexts/workflow-storage-context/types";
+import { Toast } from "../Feedback";
+import { TextField } from "../Input";
+import { client } from "../Network";
+import type { ClutchError } from "../Network/errors";
+import styled from "../styled";
+
+import { workflowByRoute } from "./utils";
+
+const Grow = styled(MuiGrow)((props: { placement: string }) => ({
+  transformOrigin: props.placement,
+}));
+
+const Popper = styled(MuiPopper)({
+  padding: "0 12px",
+  marginLeft: "12px",
+  zIndex: 1201,
+});
+
+const Paper = styled(MuiPaper)({
+  width: "400px",
+  height: "100px",
+  padding: "15px",
+  boxShadow: "0px 15px 35px rgba(53, 72, 212, 0.2)",
+  borderRadius: "8px",
+});
+
+const StyledLinkIcon = styled(IconButton)<{ $open: boolean }>(
+  {
+    color: "#ffffff",
+    marginRight: "8px",
+    padding: "12px",
+    "&:hover": {
+      background: "#2d3db4",
+    },
+    "&:active": {
+      background: "#2938a5",
+    },
+  },
+  props => ({
+    background: props.$open ? "#2d3db4" : "unset",
+  })
+);
+
+/**
+ * Component that will display a Button to generate a short link
+ * On click the component will do the following:
+ * - Read in the data from the temporary storage in the WorkflowStorageContext
+ * - Rotate it into a readable format for the API
+ * - Send an API request asking for it to be stored
+ * - If successful,
+ * -     Will switch to a readable only input field with a clipboard button to copy the short link
+ *       (this is the only time it is ever displayed)
+ * - If not successful,
+ * -     Will display a toast error message for the user
+ */
+const ShortLinker = () => {
+  const { workflows } = useAppContext();
+  const { removeWorkflowSession, retrieveWorkflowSession } = useShortLinkContext();
+  const [open, setOpen] = React.useState(false);
+  const [shortLink, setShortLink] = React.useState<string | null>(null);
+  const [validWorkflow, setValidWorkflow] = React.useState<boolean>(false);
+  const [error, setError] = React.useState<ClutchError | null>(null);
+  const anchorRef = React.useRef(null);
+  const location = useLocation();
+
+  const checkValidWorkflow = () => {
+    const workflow = workflowByRoute(workflows, location.pathname);
+
+    setValidWorkflow(workflow?.shortLink ?? false);
+  };
+
+  // will trigger on a location change, emptying out our temporary storage and rechecking the workflow
+  React.useEffect(() => {
+    if (workflows.length) {
+      checkValidWorkflow();
+    }
+
+    removeWorkflowSession();
+  }, [location]);
+
+  // Used for initial load to verify that once our workflows have loaded we are on a valid workflow
+  React.useEffect(() => {
+    if (workflows.length) {
+      checkValidWorkflow();
+    }
+  }, [workflows]);
+
+  const handleToggle = () => {
+    setOpen(!open);
```

Comment:
you'll want something like
```tsx
setOpen(o => !o);
```

Reply:
Can do

Reply:
So looking at the other poppers we use, it looks like they all just do `setOpen(!open)` as well (which is where I took it from I think)

Reply:
Will update, I'll leave the others alone for now. I eventually want to combine all these header components into a single parent component and reduce some of the duplication.